core/src/main/scala/kafka/metrics/LinuxIoMetricsCollector.scala

```diff
@@ -0,0 +1,96 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.server
+
+import java.nio.file.{Files, Paths}
+
+import org.apache.kafka.common.utils.Time
+import org.slf4j.Logger
+
+import scala.jdk.CollectionConverters._
+
+/**
+ * Retrieves Linux /proc/self/io metrics.
+ */
+class LinuxIoMetricsCollector(val procPath: String, val time: Time, val logger: Logger) {
+  import LinuxIoMetricsCollector._
+  var lastUpdateMs = -1L
+  var cachedReadBytes = 0L
+  var cachedWriteBytes = 0L
+
+  def readBytes(): Long = this.synchronized {
+    val curMs = time.milliseconds()
+    if (curMs != lastUpdateMs) {
+      updateValues(curMs)
+    }
+    cachedReadBytes
+  }
+
+  def writeBytes(): Long = this.synchronized {
+    val curMs = time.milliseconds()
+    if (curMs != lastUpdateMs) {
+      updateValues(curMs)
+    }
+    cachedWriteBytes
+  }
+
+  /**
+   * Read /proc/self/io.
+   *
+   * Generally, each line in this file contains a prefix followed by a colon and a number.
+   *
+   * For example, it might contain this:
+   * rchar: 4052
+   * wchar: 0
+   * syscr: 13
+   * syscw: 0
+   * read_bytes: 0
+   * write_bytes: 0
+   * cancelled_write_bytes: 0
+   */
+  def updateValues(now: Long): Boolean = this.synchronized {
```

Comment:
Nit: do you think the lock should be hold while reading `/proc`, or restricted to the update of `lastUpdateMs`, `cachedReadBytes` and `cachedWriteBytes`?

Reply:
Unless we choose to read this file in a background thread, there isn't a reason to avoid using a lock here.

Reply:
What I meant is not avoid the usage of a lock but restrict its scope to put it outside of the invocation of `Files#readline`.

The motivation for this is in case of I/O stall, that invocation may block for a long time (until the kernel times the operation out), with consequences for other pretendant to acquire the lock.

That said the invocation is on the `readBytes`/`writeBytes` metrics path, so *in fine* it doesn't matter.

Reply:
Assuming there's no background refresh thread, the only thread that is reading from `/proc` is the thread calling `readBytes`.  So you still have to wait for that read, whether or not you use a lock here.

Reply:
Also there is not going to be an I/O stall reading from `/proc` since `/proc` is not a real disk.